cmake: add FP16_USE_SYSTEM_* options; support find_package(FP16)

ConnorBaker opened the pull request:
Similar to the work done in https://github.com/pytorch/pytorch/pull/37137, this adds the following CMake options:

- `FP16_USE_SYSTEM_LIBS`
- `FP16_USE_SYSTEM_GOOGLEBENCHMARK`
- `FP16_USE_SYSTEM_GOOGLETEST`

This is particularly useful in the context of Nix, where we can build these libraries once and then re-use them elsewhere to avoid rebuilding vendors dependencies.

Additionally, adds a CMake configuration file to make it possible for CMake-based projects to use `find_package` to consume this library.

ConnorBaker committed:
cmake: add FP16_USE_SYSTEM_* options; support find_package(FP16)

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Updated to reflect comment from https://github.com/pytorch/cpuinfo/pull/153#discussion_r1254647992.

@Maratyszcza if you have the time, would you mind reviewing this?

meerfrau commented:
@Maratyszcza This would make the life of barebone packagers much easier. Please merge it!

BwL1289 commented:
@ConnorBaker any chance you can give me read/write permissions to the repo? would love to sync upstream

BwL1289 commented:
@ConnorBaker thanks! Looks like I need permission to push to your branch
<img width="295" height="129" alt="Screenshot 2025-08-20 at 8 33 04 AM" src="https://github.com/user-attachments/assets/cb26e9a6-f01e-446f-a631-c6863f178727" />

ConnorBaker commented:
@BwL1289 I saw the same error when viewing your PR; not sure why that came up. I've rebased on the latest and force-pushed and changed the names to have a `FP16_` prefix.

BwL1289 commented:
Thanks @ConnorBaker!
